File: src/javascript/lib/core/store.js
import Core from '../core';

function get_key(key) {
  let real_key = key;

  if (Core.global.__elixirscript_names__.has(key)) {
    real_key = Core.global.__elixirscript_names__.get(key);
  }

  if (Core.global.__elixirscript_store__.has(real_key)) {
    return real_key;
  }

  throw new Error(`Key ${real_key} not found`);
}

function create(value, name = null) {
  const key = new Core.PID();

  if (name !== null) {
    Core.global.__elixirscript_names__.set(name, key);
  }

  return Core.global.__elixirscript_store__.set(key, value);
}

function update(key, value) {
  const real_key = get_key(key);
  return Core.global.__elixirscript_store__.set(real_key, value);
}

function read(key) {
  const real_key = get_key(key);
  return Core.global.__elixirscript_store__.get(real_key);
}

function remove(key) {
  const real_key = get_key(key);
  return Core.global.__elixirscript_store__.delete(real_key);
}

export default {
  create,
  update,
  read,
  remove,
};

File: priv/std_lib/store.ex
defmodule ElixirScript.Store do

  defp get_key(key) do

    real_key = case JS.__elixirscript_names__.has(key) do
                 true ->
                   JS.__elixirscript_names__.get(key)
                 false ->
                   key
               end


    case JS.__elixirscript_store__.has(real_key) do
      true ->
        real_key
      false ->
        JS.throw JS.new(JS.Error, ["Key Not Found"])
    end
  end

  def create(key, value, name \\ nil) do
    if name != nil do
      JS.__elixirscript_names__.set(name, key)
    end

    JS.__elixirscript_store__.set(key, value)
  end

  def update(key, value) do
    real_key = get_key(key)
    JS.__elixirscript_store__.set(real_key, value)
  end

  def read(key) do
    real_key = get_key(key)
    JS.__elixirscript_store__.get(real_key)
  end

  def remove(key) do
    real_key = get_key(key)
    JS.__elixirscript_store__.delete(real_key)
  end


end
